pkg/controllers/user/pipeline/hooks/drivers/gitlab.go

```diff
@@ -0,0 +1,108 @@
+package drivers
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"github.com/xanzy/go-gitlab"
+	"io/ioutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"net/http"
+	"strings"
+)
+
+const (
+	GitlabWebhookHeader = "X-Gitlab-Event"
+	gitlabPushEvent     = "Push Hook"
+	gitlabMREvent       = "Merge Request Hook"
+	gitlabTagEvent      = "Tag Push Hook"
+)
+
+type GitlabDriver struct {
+	Pipelines          v3.PipelineInterface
+	PipelineExecutions v3.PipelineExecutionInterface
+}
+
+func (g GitlabDriver) Execute(req *http.Request) (int, error) {
+	var signature string
+	if signature = req.Header.Get("X-Gitlab-Token"); len(signature) == 0 {
+		return http.StatusUnprocessableEntity, errors.New("gitlab webhook missing token")
+	}
+	event := req.Header.Get(GitlabWebhookHeader)
+	if event != gitlabPushEvent && event != gitlabMREvent && event != gitlabTagEvent {
+		return http.StatusUnprocessableEntity, fmt.Errorf("not trigger for event:%s", event)
+	}
+
+	pipelineID := req.URL.Query().Get("pipelineId")
+	ns, id := ref.Parse(pipelineID)
+	pipeline, err := g.Pipelines.GetNamespaced(ns, id, metav1.GetOptions{})
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return http.StatusUnprocessableEntity, err
+	}
+	if pipeline.Status.Token != signature {
+		return http.StatusUnprocessableEntity, errors.New("gitlab webhook invalid token")
+	}
+
+	if pipeline.Status.PipelineState == "inactive" {
+		return http.StatusUnavailableForLegalReasons, errors.New("pipeline is not active")
+	}
+
+	if (event == gitlabPushEvent && !pipeline.Spec.TriggerWebhookPush) ||
+		(event == gitlabMREvent && !pipeline.Spec.TriggerWebhookPr) ||
+		(event == gitlabTagEvent && !pipeline.Spec.TriggerWebhookTag) {
+		return http.StatusUnavailableForLegalReasons, fmt.Errorf("trigger for event '%s' is disabled", event)
+	}
+
+	if err := utils.ValidPipelineSpec(pipeline.Spec); err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	ref, branch, commit := "", "", ""
+	envVars := map[string]string{}
+	if event == gitlabPushEvent {
+		payload := &gitlab.PushEvent{}
```

Comment:
can we perhaps make the logic prettier? By:
* creating a wrapper object with ref, branch, commit, envVars as fields
* Create separate methods for `create<Object>GitlabPushEvent`, `create<Object> gitlabMREvent `, `create<Object> gitlabTagEvent `

what do you think?